node/cn/peer.go

```diff
@@ -965,30 +965,42 @@ func (p *multiChannelPeer) UpdateRWImplementationVersion() {
 
 func (p *multiChannelPeer) ReadMsg(rw p2p.MsgReadWriter, connectionOrder int, errCh chan<- error, wg *sync.WaitGroup, closed <-chan struct{}) {
 	defer wg.Done()
-	for {
-		msg, err := rw.ReadMsg()
-		if err != nil {
-			p.GetP2PPeer().Log().Warn("ProtocolManager failed to read msg", "err", err)
-			errCh <- err
-			return
-		}
 
-		msgCh, err := p.chMgr.GetChannelWithMsgCode(connectionOrder, msg.Code)
-		if err != nil {
-			p.GetP2PPeer().Log().Warn("ProtocolManager failed to get msg channel", "err", err)
-			errCh <- err
-			return
-		}
+	var (
+		readMsgCh     = make(chan p2p.Msg, channelSizePerPeer)
+		readLoopEndCh = make(chan struct{})
+	)
 
-		if msg.Size > ProtocolMaxMsgSize {
-			err := errResp(ErrMsgTooLarge, "%v > %v", msg.Size, ProtocolMaxMsgSize)
-			p.GetP2PPeer().Log().Warn("ProtocolManager over max msg size", "err", err)
-			errCh <- err
-			return
+	go func() {
+		for {
+			msg, err := rw.ReadMsg()
```

Comment:
If `rw.ReadMsg()` hangs, then the goroutine never returns.
How about adding a timeout before calling `rw.ReadMsg()` in the old code?

Reply:
@ian0371 thanks, ian. I think about your comment a lot, but I can't imagine well.. It is hard to think up the possible case of passing through `rw.ReadMsg` with timer. 

First of all,`rw.ReadMsg` hanging will not be happened because it returns err as soon as the peer is disconnected. I have tested several times, and it always return the error.

Even assumed that the ReadMsg hangs, it has it's own timer which is set to 30s, so it can't be hung forever.
https://github.com/klaytn/klaytn/blob/0326446c5273f50aa4cd3e1ddb03d755c48d2310/networks/p2p/rlpx.go#L101

btw, this PR contains the change about avoiding the unnecessary wating when error is not related to the timeout(ex. if timeout, err would be EOF).

@hyunsooda I think this can be the answer to your comment, too. thanks for the great advice, and Is there any other concern? btw, sorry for the late answer.

Reply:
To sum up,
1. rw.ReadMsg() waits at most 30 seconds
2. If the channel is closed, it waits until the timeout
3. However, if the channel is closed, there's no need to wait. We want to execute `case <-closed` right away.

This seems like a clever way to solve the problem. What do you think @hyunsooda ?

Reply:
Or is this problematic?

```diff
diff --git a/node/cn/peer.go b/node/cn/peer.go
index f0983313..271199ae 100644
--- a/node/cn/peer.go
+++ b/node/cn/peer.go
@@ -966,6 +966,12 @@ func (p *multiChannelPeer) UpdateRWImplementationVersion() {
 func (p *multiChannelPeer) ReadMsg(rw p2p.MsgReadWriter, connectionOrder int, errCh chan<- error, wg *sync.WaitGroup, closed <-chan struct{}) {
        defer wg.Done()
        for {
+               select {
+               case <-closed:
+                       return
+               default:
+               }
+
                msg, err := rw.ReadMsg()
                if err != nil {
                        p.GetP2PPeer().Log().Warn("ProtocolManager failed to read msg", "err", err)
@@ -989,8 +995,6 @@ func (p *multiChannelPeer) ReadMsg(rw p2p.MsgReadWriter, connectionOrder int, er

                select {
                case msgCh <- msg:
-               case <-closed:
-                       return
                }
        }
 }
```

Reply:
I have tested with the above code, and it is hung at `case <-closed:` because it waits for the closed channel.